connector/connector_uaa.go

```diff
@@ -0,0 +1,158 @@
+)
+
+const (
+	UAAConnectorType = "uaa"
+)
+
+type UAAConnectorConfig struct {
+	ID           string `json:"id"`
+	ClientID     string `json:"clientID"`
+	ClientSecret string `json:"clientSecret"`
+	ServerURL    string `json:"serverURL"`
+}
+
+// standard error form returned by UAA
+type uaaError struct {
+	ErrorDescription string `json:"error_description"`
+	ErrorType        string `json:"error"`
+}
+
+type uaaOAuth2Connector struct {
+	clientID     string
+	clientSecret string
+	client       *oauth2.Client
+	uaaBaseURL   *url.URL
+}
+
+func init() {
+	RegisterConnectorConfigType(UAAConnectorType, func() ConnectorConfig { return &UAAConnectorConfig{} })
+}
+
+func (cfg *UAAConnectorConfig) ConnectorID() string {
+	return cfg.ID
+}
+
+func (cfg *UAAConnectorConfig) ConnectorType() string {
+	return UAAConnectorType
+}
+
+func (cfg *UAAConnectorConfig) Connector(ns url.URL, lf oidc.LoginFunc, tpls *template.Template) (Connector, error) {
+	uaaBaseURL, err := url.ParseRequestURI(cfg.ServerURL)
+	if err != nil {
+		return nil, fmt.Errorf("Invalid configuration. UAA URL is invalid: %v", err)
+	}
+	if !uaaBaseURL.IsAbs() {
+		return nil, fmt.Errorf("Invalid configuration. UAA URL must be absolute")
+	}
+	ns.Path = path.Join(ns.Path, httpPathCallback)
+	oauth2Conn, err := newUAAConnector(cfg, uaaBaseURL, ns.String())
+	if err != nil {
+		return nil, err
+	}
+	return &OAuth2Connector{
+		id:        cfg.ID,
+		loginFunc: lf,
+		cbURL:     ns,
+		conn:      oauth2Conn,
+	}, nil
+}
+
+func (err uaaError) Error() string {
+	return fmt.Sprintf("uaa (%s): %s", err.ErrorType, err.ErrorDescription)
+}
+
+func (c *uaaOAuth2Connector) Client() *oauth2.Client {
+	return c.client
+}
+
+func (c *uaaOAuth2Connector) Healthy() error {
+	return nil
+}
+
+func (c *uaaOAuth2Connector) Identity(cli chttp.Client) (oidc.Identity, error) {
+	uaaUserInfoURL := *c.uaaBaseURL
+	uaaUserInfoURL.Path = path.Join(uaaUserInfoURL.Path, "/userinfo")
+	req, err := http.NewRequest("GET", uaaUserInfoURL.String(), nil)
+	if err != nil {
+		return oidc.Identity{}, err
+	}
+	resp, err := cli.Do(req)
+	if err != nil {
+		return oidc.Identity{}, fmt.Errorf("get: %v", err)
+	}
+	defer resp.Body.Close()
+	switch {
+	case resp.StatusCode >= 400 && resp.StatusCode < 600:
+		// attempt to decode error from UAA
+		var authErr uaaError
+		if err := json.NewDecoder(resp.Body).Decode(&authErr); err != nil {
+			return oidc.Identity{}, oauth2.NewError(oauth2.ErrorAccessDenied)
+		}
+		return oidc.Identity{}, authErr
+	case resp.StatusCode == http.StatusOK:
+	default:
+		return oidc.Identity{}, fmt.Errorf("unexpected status from providor %s", resp.Status)
+	}
+	var user struct {
+		UserID   string `json:"user_id"`
+		Email    string `json:"email"`
+		Name     string `json:"name"`
+		UserName string `json:"user_name"`
+	}
+	if err := json.NewDecoder(resp.Body).Decode(&user); err != nil {
+		return oidc.Identity{}, fmt.Errorf("getting user info: %v", err)
+	}
+	name := user.Name
```

Comment:
The JWT token has `name` set to `""`.  So this isn't the UAA Connector failing to process the user information, it's the UAA server not setting the name properly in the token.  This could be a misconfiguration of the built-in OAuth2 client application we're using to test.  I'll keep digging.